File: agent.js
const { Configuration, OpenAIApi } = require("openai");
const { PolywrapClient } = require("@polywrap/client-js");

const readline = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

const configuration = new Configuration({
  apiKey: process.env.OPENAI_API_KEY,
});
const openai = new OpenAIApi(configuration);
const client = new PolywrapClient();

uri_allowlist = [
  "ens/wraps.eth:ens-text-record-uri-resolver-ext@1.0.0",
  "ens/wraps.eth:logger@1.0.0",
  "ens/wraps.eth:uniswap-v3@1.0.0",
];

const functions = {
  GetToolLibrary: () => new Date().toLocaleString(),
  GetFunctionsfromTool: async () => {
    const resolutionResult = await client.invoke({
      uri: "ens/wraps.eth:ens-text-record-uri-resolver-ext@1.0.0",
      method: "tryResolveUri",
      args: { authority: "ens", path: "uniswap.wraps.eth:v3" },
    });
    console.log(
      resolutionResult.ok ? resolutionResult.value : resolutionResult.error
    );
  },
  InvokeWrap: (options) => {
    console.log("Invoking wrap");
    console.log(options);
    client.invoke(options).then(console.log);
    console.log("Invoked wrap");
  },
};

const functions_description = [
  {
    name: "GetToolLibrary",
    description: "A function to get the current date and time",
    parameters: { type: "object", properties: {} },
  },
  {
    name: "GetFunctionsfromTool",
    description:
      "A function to log something in the CLI, like a console log or print",
    parameters: { type: "object", properties: {} },
  },
  {
    name: "InvokeWrap",
    description: "A function to get the ENS record",
    parameters: {
      type: "object",
      properties: {},
    },
  },
];

async function createChatCompletion(userInput, chatHistory = []) {
  chatHistory.push(
    {
      role: "system",
      content:
        "You are the Polywrap Agent. Capable of interacting with any tool, service, or API that has a wrap hosted on the Polywrap Registry.",
    },
    { role: "user", content: userInput }
  );

  const completion = await openai.createChatCompletion({
    model: "gpt-3.5-turbo-0613",
    messages: chatHistory,
    functions: functions_description,
    function_call: "auto",
  });

  let message = completion.data.choices[0].message;
  if (message.function_call) {
    console.log("-> Using a function call now");
    let functionName = message.function_call.name;
    let functionArgsString = message.function_call.arguments;
    let functionArgs = JSON.parse(functionArgsString);

    try {
      let functionResponse = functions[functionName]();
      let updatedChatHistory = [
        { role: "assistant", content: String(functionResponse) },
      ];
      console.log(functionResponse);
      chatHistory.push(...updatedChatHistory);
    } catch (error) {
      let errorMessage = String(error.message);
      let updatedChatHistory = [
        { role: "assistant", content: `Error: ${errorMessage}` },
      ];
      chatHistory.push(...updatedChatHistory);
    }
  } else {
    console.log("-> No function call used");
    console.log("Assistant: ", message.content);
    chatHistory.push({ role: "assistant", content: message.content });
  }

  continueChat(chatHistory);
}

function continueChat(chatHistory) {
  readline.question("Human Feedback: ", (userInput) => {
    createChatCompletion(userInput, chatHistory).catch((error) =>
      console.error(error.message)
    );
    console.log(chatHistory);
  });
}

continueChat([]);

File: agent.py
import openai
import json
import os 

from dotenv import load_dotenv
from polywrap_client import PolywrapClient
from typing import cast

from pathlib import Path
from polywrap_client import PolywrapClient, ClientConfig
from polywrap_core import Uri, InvokerOptions, UriPackageOrWrapper, UriResolver
from polywrap_client_config_builder import PolywrapClientConfigBuilder
from polywrap_uri_resolvers import FsUriResolver,SimpleFileReader, StaticResolver, RecursiveResolver
from polywrap_uri_resolvers import UriResolverAggregator
from polywrap_http_plugin import http_plugin


import asyncio

import nest_asyncio

nest_asyncio.apply()

load_dotenv()

ipfs_wrapper_path = Path("fs//Users/robertohenriquez/pycode/polywrap/hackathon/Auto-GPT/autogpt/auto_gpt_workspace/wrappers/ipfs-http-client")


resolver = RecursiveResolver(
        UriResolverAggregator(
            [
                cast(UriResolver, FsUriResolver(file_reader=SimpleFileReader())),
                cast(UriResolver, StaticResolver({Uri("wrap://ens/wraps.eth:http@1.1.0", ipfs_wrapper_path): http_plugin()})),
            ]
        )
    )
config = ClientConfig(resolver=resolver)
client = PolywrapClient(config)

openai.api_key = os.getenv("OPENAI_API_KEY")


def polywrap_bignumber(base_number, factor):
    """Multiplies a base number by a factor to get a new number"""
    uri = Uri.from_str(
        f'fs//Users/robertohenriquez/pycode/cloud/AGI/Auto-GPT/autogpt/auto_gpt_workspace/wrappers/bignumber'
    )
    args = {
        "arg1": str(base_number),  # The base number
        "obj": {
            "prop1": str(factor),  # multiply the base number by this factor
        },
    }
    options: InvokerOptions[UriPackageOrWrapper] = InvokerOptions(
        uri=uri, method="method", args=args, encode_result=False
    )
    print(asyncio.run(client.invoke(options)))
    result = asyncio.run(client.invoke(options))
    return f"the result is {result}"

def fetch_tool_library(tool_name):
    """a mapping of keywords to tools that ChatGPT will be able to use through the polywrap python client"""
    # Mapping of keywords to tools
    tool_mappings = {
        "ipfs": ["wrap/ipfs"],
        "http": ["wrap/http"],
        "ens": ["wrap/ens"],
        "ethers": ["wrap/ethers"],
        "ethereum": ["wrap/ethers"],
    }
    
    # Check if the tool_name has any keywords
    keywords = tool_name.lower().split()
    matching_tools = []
    
    for keyword in keywords:
        if keyword in tool_mappings:
            matching_tools.extend(tool_mappings[keyword])
    
    # Remove duplicates and sort the tools alphabetically
    matching_tools = sorted(list(set(matching_tools)))
    
    return str(matching_tools)

def invoke_tool(options):
    """The invoker function that Chat GPT uses to invoke tools with polywrap"""
    
    uri = Uri.from_str(options["tool_uri"])
    args = options["arguments"]
    method = options["method"]
    options: InvokerOptions[UriPackageOrWrapper] = InvokerOptions(
        uri=uri, method=method, args=args, encode_result=False
    )
    print(asyncio.run(client.invoke(options)))
    result = asyncio.run(client.invoke(options))
    return f"SUCCESS! The result is: {result}"



question = "Find what tools are available for ipfs"
# Define the functions for our ChatGPT Agent

functions = [
    {
        "name": "polywrap_bignumber",
        "description": "Use this function every time you need to make a multiplication",
        "parameters": {
            "type": "object",
            "properties": {
                "base_number": {
                    "type": "number",
                    "description": "The base number",
                },
                "factor": {
                    "type": "number",
                    "description": "The factor to multiply the base number",
                },
            },
            "required": ["base_number", "factor"],
        },
    },
    {
        "name": "fetch_tool_library",
        "description": "Fetches information about a specific tool from the tool library, given the tool name.",
        "parameters": {
            "type": "object",
            "properties": {
                "tool_name": {
                    "type": "string",
                    "description": "The name of the tool to fetch",
                },
            },
            "required": ["tool_name"],
        },
    },
    {
        "name": "invoke_tool",
        "description": "Invokes the given tool with the specified arguments",
        "parameters": {
            "type": "object",
            "properties": {
                "options": {
                    "type": "object",
                    "description": "The options to pass to the tool, including the tool_uri and the arguments",
                },
            },
            "required": ["tool_uri", "options"],
        },
    },
]






def agent_loop(question, functions, chat_history=None):
    """
    A simple agent loop that can be used to continue a conversation with ChatGPT, using the polywrap python client to invoke tools
    
    Args:
        question (str): The question to ask ChatGPT
        functions (list): The list of functions to use with ChatGPT 
        chat_history (list): The chat history to use with ChatGPT
    
    Returns:
        response (dict): The response from ChatGPT which sometimes is empty #TODO fix this
        chat_history (list): The updated chat history to be used in the follow up call to the agent loop
    """
    if chat_history is None:
        chat_history = []

    chat_history.append({"role": "user", "content": question})

    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo-0613",
        messages=chat_history,
        functions=functions,
        function_call="auto",
    )

    message = response["choices"][0]["message"]

    if message.get("function_call"):
        print("-> Using a function call now")
        function_name = message["function_call"]["name"]
        function_args_string = message["function_call"]["arguments"]
        function_args = json.loads(function_args_string)

        try:
            function_response = globals()[function_name](**function_args)
            updated_chat_history = [{"role": "assistant", "content": function_response}]
            chat_history.extend(updated_chat_history)
        except Exception as e:
            error_message = str(e)
            updated_chat_history = [{"role": "assistant", "content": f"Error: {error_message}"}]
            chat_history.extend(updated_chat_history)
    else:
        print("-> No function call used")
        chat_history.append({"role": "assistant", "content": message["content"]})
        return response, chat_history

    return response, chat_history


# Example usage
question = "Find what tools are available for ipfs in the skill library"
response, chat_history = agent_loop(question, functions)

print(chat_history)

# Continue the conversation
new_question = "Call the function invoke_tool to the received URI with the arguments: {method: get, arguments: {hash: 'Qm11234'}, tool_uri: <the found URI>}"
response, chat_history = agent_loop(new_question, functions, chat_history)

# Continue the conversation further if needed
print(chat_history, response)

# Example usage
for i in range(2):
    new_question = "Try to call the invoke_tool function again with the provided feedback until you get the success result"
    response, chat_history = agent_loop(new_question, functions, chat_history)

print(chat_history)


